### PR TITLE
Light Cycles

### PR DESCRIPTION
**Pull request by Shadowmech88**
http://puu.sh/t9e6B/c8ff166839.webm

Adds light rods.
Adds light cycles.

Light rods are assigned a random color when they spawn, or can be given a custom color as an argument to their `New()`. When used, a light rod manifests a light cycle around the user. When used again, the light cycle disappears.
Light cycles move forward constantly while they have an occupant. The occupant can change the light cycle's direction, but not directly influence its movement. When a light cycle bumps into any dense object, the occupant is gibbed and the light cycle disappears. Light cycles pass through doors and mobs without bumping them. Light cycles leave a trail of light ribbon objects as they move. These can be removed by attacking them with any object, as long as you're not riding a light cycle.

I figure these could be a neat tool for adminbus-games in the same vein as the Bomberman equipment.

### CODE REVIEW

**Comment by Probe1** [comment]
siiiick

How are they deleted?

**Comment by Shadowmech88** [comment]
What do you mean? Like when they crash?
They just get qdel()'d.
The rider drops the light rod they were using so it can be picked up and used to summon the cycle again.

The cycles are really only supposed to exist when they have someone riding them.

**Comment by Intigracy** [comment]
He means the light barriers.

**Comment by Shadowmech88** [comment]
Oh, you just smack the barrier with something to make it disappear from that tile.

**Comment by 9600bauds** [comment]
Maybe hitting a color of barrier with that colors' corresponding light rod deletes all light barriers of that color?

File: code/game/objects/structures/vehicles/lightcycle.dm [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+		dir = direction
+		user.dir = direction
+
+	update_mob()
+
+/obj/structure/bed/chair/vehicle/lightcycle/Bump(atom/A)
+	if(istype(A, /mob))
+		var/mob/M = A
+		src.visible_message("<span class='warning'>\The [occupant] knocks over \the [M] with \his [src.name]!</span>")
+		M.stop_pulling()
```
Comment (9600bauds) [inline code/game/objects/structures/vehicles/lightcycle.dm]: I believe this is implicit for Knockdown() and not needed but whatever

File: code/game/objects/structures/vehicles/lightcycle.dm [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+		M.Stun(8)
+		M.Knockdown(5)
+		M.lying = 1
```
Comment (9600bauds) [inline code/game/objects/structures/vehicles/lightcycle.dm]: This is DEFINITELY implicit for Knockdown() and not needed

File: code/game/objects/structures/vehicles/lightcycle.dm (hunk 1 above)
Comment (Shadowmech88) [reply code/game/objects/structures/vehicles/lightcycle.dm]: I took it from the code executed when a MULE runs over a mob, I don't know how old it is.

Comment (9600bauds) [reply code/game/objects/structures/vehicles/lightcycle.dm]: hahaha oh god

**Comment by Shadowmech88** [comment]
I suppose I could add that, but why? There's no delay on attacking the barriers, so clearing them out isn't much of a task.

**Comment by 9600bauds** [comment]
Oh there's no delay, neat
I figured to quickly clear the board during adminbus

**Comment by Shadowmech88** [comment]
I'd have to keep a global list of lists of all the light walls of each color, right?
Or would it be better to just have each light wall check for adjacent ones of the same color and delete them when it's deleted, so that the effect propagates outwards along the wall?

File: code/game/objects/structures/vehicles/lightcycle.dm (hunk 1 above)
Comment (Shadowmech88) [reply code/game/objects/structures/vehicles/lightcycle.dm]: Think I should go ahead and remove it?

**Comment by 9600bauds** [comment]
Well you could just keep the list in the rod itself, if you say cleaning these up isn't a hassle though it's okay

**Comment by Shadowmech88** [comment]
I was thinking of the hassle for other players after lightcycles ride through the station, not for the admins in clearing the game board. Something to clean it up all at once would probably be good.

**Comment by Shadowmech88** [comment]
I'll make it a proc that admins can call manually so riders can't use their rods to instantly reclaim a bunch of space for themselves.

**Comment by 9600bauds** [comment]
If these cycles are meant to be used inside the station at all, consider giving them PASSDOOR so they can at least move through the hallways/firelocks

**Comment by Aurx** [comment]
Admins already have tools to mass-delete objs, and it'd be easy to use SDQL to wipe out all lightribbons of a certain color.  I'm not sure how useful another mass-delete proc for just lightribbons would be.

**Comment by Exxion** [comment]
There's a goddamn delete all of type option in VV, if memory serves

**Comment by ghost** [comment]
This is something I've wanted in SS13 for the longest time. I am damn impressed.

**Comment by Aurx** [comment]
A proc that has to be manually called via debug tools is going to be forgotten by or never known to admins.

**Comment by Shadowmech88** [comment]
The ones who don't know can still run around the arena swinging a crowbar to take the light ribbons down.

Do you have a better suggestion for accessing the deletion proc?

**Comment by Aurx** [comment]
Would you be able to make it so that an adminghost clicking a lightribbon opens a  "Delete this ribbon Y/N?" prompt and nukes it if Y?  I think that'd be the most elegant solution because it wouldn't need extra verbs.  Fuck if I know UI practices though, the codermins would likely have the best input.

File: code/game/objects/structures/vehicles/lightcycle.dm [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+	..(T)
+	var/hue = rand(0,360)
+	if(col)
+		var/red = GetRedPart(col)
+		var/green = GetGreenPart(col)
+		var/blue = GetBluePart(col)
+		var/list/hsl = rgb2hsl(red,green,blue)
+		hue = hsl[1]
+	var/list/rgb_list = hsl2rgb(hue, 100, 50)		//only neon colors
+	l_color = "[rgb(rgb_list[1],rgb_list[2],rgb_list[3])]"
```
Comment (PJB3005) [inline code/game/objects/structures/vehicles/lightcycle.dm]: `rgb` already returns a string, so you don't need the wrapping string.

File: code/game/objects/structures/vehicles/lightcycle.dm [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+	if(cycle_active)
+		icon_state = "lightcycle_keys_active"
+		cant_drop = 1
+	else
+		icon_state = "lightcycle_keys_inactive"
+		cant_drop = 0
+
+/obj/item/key/lightcycle/attack_self(mob/user)
+	..()
+	if(!istype(user, /mob/living/carbon/human))
```
Comment (PJB3005) [inline code/game/objects/structures/vehicles/lightcycle.dm]: `ishuman`

File: code/game/objects/structures/vehicles/lightcycle.dm [hunk 5]
```diff
@@ -0,0 +1,252 @@
+/obj/item/key/lightcycle
+	name = "light rod"
+	desc = "A strange-looking glowing rod."
+	icon = 'icons/obj/vehicles.dmi'
+	icon_state = "lightcycle_keys_inactive"
+	w_class = W_CLASS_TINY
+	var/cycle_active = FALSE
+	var/obj/structure/bed/chair/vehicle/lightcycle/summoned_cycle = null
```
Comment (PJB3005) [inline code/game/objects/structures/vehicles/lightcycle.dm:8]: Any reason you can't just replace any code checking `cycle_active` with whether or not `summoned_cycle` is null or not?

Comment (PJB3005) [reply code/game/objects/structures/vehicles/lightcycle.dm:8]: Also, the vehicle code for both keys and vehicles themselves track eachother already, any reason you can't use the key's paired vehicle variable and vice versa?

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/vehicles/lightcycle.dm (hunk 5 above)
Comment (Shadowmech88) [reply code/game/objects/structures/vehicles/lightcycle.dm:8]: Can I re-cast vars as different types from their parent? The `paired_to` var of keys is cast as an `/obj/structure/bed/chair/vehicle` meaning I couldn't access light cycle-specific vars and procs.

Comment (PJB3005) [reply code/game/objects/structures/vehicles/lightcycle.dm:8]: Oh, right, didn't think of that.

It's fine then.

**Comment by ihadtoregisterforthis** [comment]
The trails could despawn automatically after a reasonably long time, like a minute or two. It's not like someone riding it would survive forever.

Also change light cycles to lightcycles please

**Comment by Shadowmech88** [comment]
According to https://en.wikipedia.org/wiki/Tron_(franchise) the correct phrasing is "light cycle."

**Review by PJB3005 — APPROVED** [review]

**Comment by Intigracy** [comment]
The point of the controller smacking the lights to remove them all was so that PLAYERS could remove them all easily.

**Comment by Shadowmech88** [comment]
But then players could just easily clear half the board in the middle of a duel by getting off their cycle and hitting their trail once.

**Comment by Aurx** [comment]
Now we need a verb that puts everybody on a lightcycle and doesn't let them get off.

**Comment by Shadowmech88** [comment]
I could make a child of the keys that disallows dismounting, and then the verb could just spawn a key in everyone's hand and call `attack_self()`.

**Comment by Aurx** [comment]
You'd also need to check for hand existence and forcedrop whatever they were holding, wouldn't you?

**Comment by Shadowmech88** [comment]
Correct.

**Pull request merged**